frontend/utils/faviconBadge.js

```diff
@@ -0,0 +1,253 @@
+  // as detailed here: https://github.com/mrdoob/three.js/issues/1305
+  if (!src?.match(/^data/) && options.crossOrigin) {
+    faviconImage.crossOrigin = 'anonymous'
+  }
+
+  faviconImage.src = src
+}
+
+const updateTitle = (label) => {
+  if (options.fallback) {
+    // Grab the current title that we can prefix with the label
+    let originalTitle = document.title
+
+    // Strip out the old label if there is one
+    if (originalTitle[0] === '(') {
+      originalTitle = originalTitle.slice(originalTitle.indexOf(' '))
+    }
+
+    if ((label + '').length > 0) {
+      document.title = '(' + label + ') ' + originalTitle
+    } else {
+      document.title = originalTitle
+    }
+  }
+}
+
+const drawBubble = (context, label, color) => {
+  // automatic abbreviation for long (>2 digits) numbers
+  if (typeof label === 'number' && label > 99 && options.abbreviate) {
+    label = abbreviateNumber(label)
+  }
+
+  // bubble needs to be larger for double digits
+  const len = (String(label)).length - 1
+  const width = options.width * r + (6 * r * len)
+  const height = options.height * r
+  // const top = size - height
+  // const left = size - width - r
+  const bottom = 16 * r
+  const right = 16 * r
+  // const radius = 2 * r
+  const arcRadius = Math.min(width, height) / 2
+
+  // webkit seems to render fonts lighter than firefox
+  context.font = (browser.webkit ? 'bold ' : '') + options.font
+  context.fillStyle = options.background
+  context.strokeStyle = options.background
+  context.lineWidth = r
+
+  // bubble
+  // draw circle bubble
+  context.arc(right - arcRadius, bottom - arcRadius, arcRadius, 0, 2 * Math.PI)
+
+  // draw rectangle bubble
+  // context.beginPath()
+  // context.moveTo(left + radius, top)
+  // context.quadraticCurveTo(left, top, left, top + radius)
+  // context.lineTo(left, bottom - radius)
+  // context.quadraticCurveTo(left, bottom, left + radius, bottom)
+  // context.lineTo(right - radius, bottom)
+  // context.quadraticCurveTo(right, bottom, right, bottom - radius)
+  // context.lineTo(right, top + radius)
+  // context.quadraticCurveTo(right, top, right - radius, top)
+  // context.closePath()
+  context.fill()
+
+  // bottom shadow
+  // context.beginPath()
+  // context.strokeStyle = 'rgba(0,0,0,0.3)'
+  // context.moveTo(left + radius / 2.0, bottom)
+  // context.lineTo(right - radius / 2.0, bottom)
+  // context.stroke()
+
+  // label
+  context.fillStyle = options.color
+  context.textAlign = 'right'
+  context.textBaseline = 'top'
+
+  // unfortunately webkit/mozilla are a pixel different in text positioning
+  // NOTE: temporarily comment this drawing text because we don't need it
+  // context.fillText(label, r === 2 ? 29 : 15, browser.mozilla ? 7 * r : 6 * r)
+}
+
+const refreshFavicon = () => {
+  // check support
+  if (getCanvas() instanceof HTMLCanvasElement) {
+    setFaviconTag(getCanvas().toDataURL())
+  }
+}
+
+const abbreviateNumber = (label): string => {
```

Comment:
As far as this code `FaviconBadge.setBubble(this.shouldSetBadge ? 1 : 0)` looks, currently the only possible value for `label` is either 1 or 0 and this `abbreviateNumber` function isn't doing anything. If there is no future case that this becomes useful, Can this  `abbreviateNumber` and the related logic be removed so that the file is kept concise and dry?

Reply:
I will remove this function since we are not using label at the moment and bubble is only the thing we need at the moment.